Implement the `aspect-ratio` property for replaced elements

valadaptive opened the pull request:
First step towards fully implementing `aspect-ratio`. This is probably the easy part since replaced elements are already known to have an intrinsic aspect ratio and treated as such.

This PR touches the same code as #32777 and therefore depends on it.

This will require the `aspect-ratio` property to be enabled for layout 2020 in Stylo--I currently have it pointed to my fork which does so.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] ~~These changes fix #___ (GitHub issue number if applicable)~~

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

valadaptive committed:
Implement `aspect-ratio` for replaced elements

There are two regressions because we don't implement `object-fit`, and
one because we don't properly represent non-available <img>s with `alt`
attributes.

Signed-off-by: valadaptive <valadaptive@protonmail.com>

valadaptive commented:
This still needs a bit more work--I need to enable the aspect-ratio WPT tests and fix a bug relating to max-width and border-box

valadaptive committed:
Add WPT tests for box-sizing with aspect-ratio

Signed-off-by: valadaptive <valadaptive@protonmail.com>

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#47187) with upstreamable changes.

valadaptive commented:
Turns out the "bug" was me forgetting to rebuild.

On the plus side, I did create some web platform tests (WPT tests? Has that "ATM machine" ring to it) for testing the behavior of `aspect-ratio` in combination with `box-sizing: border-box`, which seems to have been missing from the existing test suite.

Most of the existing `aspect-ratio` tests seem to be single-file ones, but there are a lot of different test cases and I don't want to create 24 individual .html files, so I used the layout test harness.

Not sure why the WPT Export step is failing--doesn't seem to be anything related to the contents of the tests themselves.

mrobinson commented:
> Not sure why the WPT Export step is failing--doesn't seem to be anything related to the contents of the tests themselves.

This might be related to the commit from #32803 somehow. Hopefully once you rebase it will improve, but I'll watch it.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#47187).

valadaptive commented:
Seems to have been caused by force-pushing while the initial WPT sync action was running; I believe it tried to create the PR twice?

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#47187).

valadaptive commented:
About the regressions: replaced-element-024 and replaced-element-026 regress because we don't implement `object-fit`, and replaced-element-028 regresses because we don't treat non-available `<img>`s with `alt` attributes properly.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#47187).

mrobinson commented on components/layout_2020/style_ext.rs:
```diff
@@ -535,6 +578,77 @@ impl ComputedValuesExt for ComputedValues {
+
+        // For all cases where a ratio is specified:
+        // "If the <ratio> is degenerate, the property instead behaves as auto."
+        if let PreferredRatio::Ratio(ratio) = preferred_ratio {
+            if ratio.is_degenerate() {
+                preferred_ratio = PreferredRatio::None;
+            }
+        }
```
```suggestion
        if matches!(preferred_ratio,  PreferredRatio::Ratio(ratio) if ratio.is_degenerate()) {
            preferred_ratio = PreferredRatio::None;
        }
```

mrobinson commented on components/layout_2020/style_ext.rs:
```diff
@@ -535,6 +578,77 @@ impl ComputedValuesExt for ComputedValues {
 
+    /// Resolve the preferred aspect ratio according to the given natural aspect
+    /// ratio and the `aspect-ratio` property.
+    /// https://drafts.csswg.org/css-sizing-4/#aspect-ratio
```
Raw URLS in rustdoc comments lead to rustdoc warnings:

```suggestion
    /// <https://drafts.csswg.org/css-sizing-4/#aspect-ratio>
```

mrobinson commented on components/layout_2020/style_ext.rs:
```diff
@@ -143,6 +145,42 @@ impl PaddingBorderMargin {
+    /// border-box, and the aspect-ratio property does not contain "auto", then
+    /// the aspect ratio is in respect to the border box. This will then contain
+    /// the summed sizes of the padding and border. Otherwise, it's 0.
+    pb_adjust: LogicalVec2<Au>,
```
Let's just calling this `box_sizing_adjustment`.

mrobinson reviewed (approved):
Excellent work. I have a few nits, but I'll apply them myself and try to land this.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#47187).
